### DISCUSSION TITLE
How would I open this .csv file?

### DISCUSSION THREAD

**Discussion by endolith** [binjr/binjr#136] [Q&A] [answered]
```
# Total Phase Data Center(tm) v7.01
# (c) 2005-2021 Total Phase, Inc.
# www.totalphase.com
#
# 
#
# Level,Sp,Index,Date h:m:s.ms.us,Dur,Len,Err,Dev,Ep,Record,Summary,Data
0,,0,May 03 16:52:24.000.000,,,,,,Capture started (Aggregate),[05/03/23 16:52:24],
0,FS,10,May 03 16:52:26.284.300,118.416 us,8 B,,00,00,Get Device Descriptor,Index=0 Length=64,12 01 00 02 00 00 00 08
0,FS,32,May 03 16:52:26.372.198,52.000 us,0 B,,00,00,Set Address,Address=28,…
```

I tried using the Edit CSV Parsing Profile but I'm not sure how to use it 

![2023-05-04 16_02_54-Edit CSV Parsing Profile](https://user-images.githubusercontent.com/58611/236316900-e514f5c5-4bd9-4ef3-b84f-02d1342cccb7.png)

and [the documentation ](https://binjr.eu/documentation/user_guide/editing_parsing_profiles/) is "under construction"

**Comment by endolith** [comment, marked as answer]
Wait I just figured it out.  `\w(3)` instead of `\w{3}`.  Then I had to remove the "TIMEZONE" thing and it parses:

![2023-05-04 16_09_16-Edit CSV Parsing Profile](https://user-images.githubusercontent.com/58611/236318347-a2230334-1e56-4568-80c0-d202c6f5caea.png)

But then it doesn't understand the other fields so I have to process the CSV file into a format it recognizes anyway, so never mind.

**Reply by fthevenet** [reply]
Hi,

There are advanced parsing functionalities available for dates, but for all other columns, it is going to assume data can be parsed as decimal numbers (or treat it as Not-a-Number)
The only parameters you can play with, is the locale used  to parse the numbers, which will determines things like what character is used as the decimal separator.

Also, you might be surprised to see all your timestamps dated from 1969: you can find a detailed explanation of why this is the case here: #129
TL;DR this is because the original data does not contain year information, so a placeholder is used (right now, 1/1/1970 UTC). The next version will let you choose between that and the current date for the placeholder.

**Reply by endolith** [reply]
I had ChatGPT convert my existing matplotlib script into multiple CSV files, and that kind of works (except I don't know how to get it to recognize microseconds):

```
Timestamp,Out
2023-05-03 16:52:28.617518,288.0
2023-05-03 16:52:28.618518,288.0
2023-05-03 16:52:28.619518,288.0
…
```

```
Timestamp,In
2023-05-03 16:52:33.964150,384.0
2023-05-03 16:52:33.965151,288.0
2023-05-03 16:52:33.966151,288.0
2023-05-03 16:52:33.967151,288.0
…
```

```
Timestamp,FB
2023-05-03 16:52:28.625718,48125.0
2023-05-03 16:52:28.633718,48125.0
2023-05-03 16:52:28.641719,48125.0
2023-05-03 16:52:28.649720,48125.0
2023-05-03 16:52:28.657712,48125.0
…
```

![2023-05-04 16_43_52-Total Phase csv](https://user-images.githubusercontent.com/58611/236325142-bd06363d-b81d-4424-aa1b-e3f2c4fe7b12.png)

![image](https://user-images.githubusercontent.com/58611/236328699-95aca2ed-01b8-44fa-81f6-30b66b414ff4.png)

And I can't load multiple CSVs at the same time, I have to use the open dialog separately for each one?  

(I'm looking for something that's faster and more interactive than matplotlib with hundreds of thousands of points.)

Is there a way to show text-labeled events as vertical lines or some other type of flag?

Also, I wish Auto Scale Y Axis didn't use 0 as the minimum, and I wish I could scroll and zoom with mouse wheel, like an audio file editor (Ocenaudio, Adobe Audition, etc.)

**Comment by fthevenet** [comment]
> *I had ChatGPT convert my existing matplotlib script into multiple CSV files, and that kind of works (except I don't know how to get it to recognize microseconds):*

Indeed, there isn't a `MICRO` predefined temporal group in the parsing settings (though there is a `NANO` one). Also, the `FRACTION` element behaviour isn't doing what it should, as is the same as `MILLI`, when it should really take a number as the decimal part of a second; I'll look into changing that. 

I'd say that right now you could multiply the number of µs by 1000 and use the `NANO` temporal group on that, but unfortunately, it's not going to be of a great help.

The bigger issue here is that the granularity for timestamps in the CSV adapter is imitated to the closest millisecond. 
The reason for that is timestamps are represented as Unix epoch milliseconds (ms since 01/01/1970), so that they can fit inside a `long` type (signed 64 bits integer) in the backend.
I'm not sure I want to change that representation for the part that is used to sort and select ranges of samples, as this could have some serious knock on effect on performance and resource usage (as well as making the implementation more complex), **but** what might still be feasible is to store the actual value for the timestamp with a higher precision so that it can be used when plotting the sample.
What this means is that while you would still be limited to the millisecond when selecting a range of values to plot, the actual samples would be drawn onto the X axis with a better accuracy (e.g. 1970/01/01T00:00:00.000100 and 1970/01/01T00:00:00.000900 would not be drawn at the same coordinate as they are now).

If you believe that would be a valuable enhancement despite its limitations, I'll look into it in more details.

**Comment by fthevenet** [comment]
> *And I can't load multiple CSVs at the same time, I have to use the open dialog separately for each one?*

In the current version, you cannot load more that one file at a time, but I agree that this seriously hampers productivity in some use cases.
As a matter of fact, this point has already been identified as a needed enhancement for the CSV adapter: https://github.com/orgs/binjr/discussions/119, but it is the only one from that list that has not been implemented so far.
I definitely expect it to be done at some point, however, given the fact that a similar mechanism already exists for another source adapter (for log files), where you can select a folder or a zip file in a single action.

**Comment by fthevenet** [comment]
> *Is there a way to show text-labeled events as vertical lines or some other type of flag?*

Adding annotations on the charts, in the form of labelled vertical markers or areas, is not supported at the moment.
Although it has been on the back of my mind for a long time, and requested by others too it would require a fairly significant effort, as a bunch of extra UI would need to be added to make it usable.
It's been repeatedly pushed down the priority list so far, but I definitely see that as a valuable addition.

**Comment by fthevenet** [comment]
> *Also, I wish Auto Scale Y Axis didn't use 0 as the minimum,* 

That's a good point, and actually something I implemented as an option but never surfaced in the UI. I'll add a toggle in the chart properties panel where the user can decide if 0 should always be forced into the auto range or not.

> *and I wish I could scroll and zoom with mouse wheel, like an audio file editor (Ocenaudio, Adobe Audition, etc.)*

Actually, you can! Press `shift` + `mouse wheel` to zoom in or out, and `alt` + `mouse wheel` to pan right or left.

**Reply by endolith** [reply]
shift + mouse wheel seems to expand the axes sizes vertically?  Not a horizontal zoom in to the data as I would expect.  alt + mouse wheel does pan, so that's good.

**Reply by fthevenet** [reply]
Yes, you're right: `ctrl` + `mouse wheel` zooms in or out on the X axis (which is what I meant to write in my answer above), but `shift` + `mouse wheel` does indeed something different and actually changes charts heights.  

This is the behaviour originally intended for that feature and not actually a bug, but I can understand why you say it is unexpected. 
In hindsight, it would certainly be less surprising if `shift` + `mouse wheel`  did a zoom on the Y axis (in the same way `ctrl` does on the X axis) and the current behaviour was assigned to a different shortcut.
